packages/core/src/http_clients/base_http_client.ts

```diff
@@ -0,0 +1,239 @@
+import type { Readable } from 'stream';
+
+import { applySearchParams, type SearchParams } from '@crawlee/utils';
+
+import type { FormDataLike } from './form_data_like';
+
+type Timeout =
+    | {
+          lookup: number;
+          connect: number;
+          secureConnect: number;
+          socket: number;
+          send: number;
+          response: number;
+      }
+    | { request: number };
+
+type Method =
+    | 'GET'
+    | 'POST'
+    | 'PUT'
+    | 'PATCH'
+    | 'HEAD'
+    | 'DELETE'
+    | 'OPTIONS'
+    | 'TRACE'
+    | 'get'
+    | 'post'
+    | 'put'
+    | 'patch'
+    | 'head'
+    | 'delete'
+    | 'options'
+    | 'trace';
+
+/**
+ * Maps permitted values of the `responseType` option on {@apilink HttpRequest} to the types that they produce.
+ */
+export interface ResponseTypes {
+    'json': unknown;
+    'text': string;
+    'buffer': Buffer;
+}
+
+interface Progress {
+    percent: number;
+    transferred: number;
+    total?: number;
+}
+
+// TODO BC with got - remove the options and callback parameters in 4.0
+interface ToughCookieJar {
+    getCookieString: ((
+        currentUrl: string,
+        options: Record<string, unknown>,
+        callback: (error: Error | null, cookies: string) => void,
+    ) => string) &
+        ((url: string, callback: (error: Error | null, cookieHeader: string) => void) => string);
+    setCookie: ((
+        cookieOrString: unknown,
+        currentUrl: string,
+        options: Record<string, unknown>,
+        callback: (error: Error | null, cookie: unknown) => void,
+    ) => void) &
+        ((rawCookie: string, url: string, callback: (error: Error | null, result: unknown) => void) => void);
+}
+
+interface PromiseCookieJar {
+    getCookieString: (url: string) => Promise<string>;
+    setCookie: (rawCookie: string, url: string) => Promise<unknown>;
+}
+
+type SimpleHeaders = Record<string, string | string[] | undefined>;
+
+/**
+ * HTTP Request as accepted by {@apilink BaseHttpClient} methods.
+ */
+export interface HttpRequest<TResponseType extends keyof ResponseTypes = 'text'> {
```

Comment:
We may not want to keep this. If people use `responseType` in current `sendRequest`, the type is incorrect anyway. And I'd guess that using e.g., `gotScraping({...}).json()` is more prevalent anyway.

Reply:
if the types were wrong before, i am fine with doing this right ahead. worst case we can fix it if someone complains

Reply:
Okay, looks like I was too optimistic, this needs to stay. `got` does have `.json`, `.text` and `.buffer` methods, but they're on the promise returned from the request function, not on the response the promise resolves to. 

I'd rather do this in a cleaner fashion (`response = await httpClient.sendRequest(); await response.json()`) than invent nonsense here.